### ISSUE TITLE
Feature Request: Implement `torch.sparse.spdiags`

### ISSUE THREAD

**Issue by krshrimali** [pytorch/pytorch#70926]
## Feature

Add `torch.sparse.spdiags` function which returns a sparse matrix from the given diagonal arrays and offsets. This is similar to SciPy's [`scipy.sparse.spdiags`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spdiags.html#scipy-sparse-spdiags).

## Motivation and Context

Please see issue: https://github.com/pytorch/pytorch/issues/47640 for motivation, and more context.

[This comment](https://github.com/pytorch/pytorch/issues/47640#issuecomment-1006541377) by @IvanYashchuk is a starting point of how the signature might look like for `torch.sparse.spdiags`:

```python
torch.sparse.spdiags(Tensor diagonals, IntTensor offsets, int[] shape, Layout? layout=None) -> SparseTensor
```

Note: The support should be for both Sparse COO and Sparse CSR layouts in PyTorch.

cc: @IvanYashchuk @pearu @mruberry

**Comment by cpuhrsch** [comment]
I think this looks good. Other PyTorch factory functions also have options for dtype, device, etc. Is that something we might want here as well?

For example

```
torch.zeros(*size, *, out=None, dtype=None, layout=torch.strided, device=None, requires_grad=False) → Tensor
```

So we could have 
```
torch.sparse.spdiags(Tensor diagonals, Tensor offsets, *size, *, out=None, dtype=None, layout=torch.coo, device=None, requires_grad=False) → Tensor
```

**Comment by krshrimali** [comment]
Good point, @cpuhrsch! IMO, yes - those arguments are something we should add here as well. These are also there in [`torch.sparse_csr_tensor`](https://pytorch.org/docs/stable/generated/torch.sparse_csr_tensor.html). I'll update the signature in the description. Thank you!

**Comment by IvanYashchuk** [comment]
It's not a factory function, it's a regular PyTorch operation with Tensor in - Tensor out. Let's take a look at similar functions from `torch` namespace:

- `torch.diagflat(input, offset=0) → Tensor`
- `torch.diag_embed(input, offset=0, dim1=- 2, dim2=- 1) → Tensor`
- `torch.diagonal(input, offset=0, dim1=0, dim2=1) → Tensor`

They take an existing Tensor and dtype, device, requires_grad should be inherited from the input. I don't think we should add `dtype=`, `device=`, `requires_grad=` here.

`torch.zeros` example doesn't take Tensors as input and that's why it requires additional info to create a Tensor.

**Comment by cpuhrsch** [comment]
@IvanYashchuk - sounds good! Does that mean we'll also add derivatives [like we do for diagonal](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L500-L502)?

**Comment by IvanYashchuk** [comment]
Yes, we should get the autograd working either by making the implementation to be a composite of existing autograd-supported operations or by adding an entry to `derivatives.yaml`.

**Comment by nabriis** [comment]
This would be a really nice feature to add!

**Comment by amjames** [comment]
@nabriis If you could provide some information about your use case. In particular is the generalization to N-D output important, and if it is,  what the generalized interface might look like. I have provided some alternatives for a generalized interface in #78439.

**Comment by amjames** [comment]
Also pinging @rjkilpatrick who opened the original feature request. Any information you have about use case would be welcome!

**Comment by vadimkantorov** [comment]
It would be nice if docs of https://pytorch.org/docs/stable/generated/torch.sparse.spdiags.html explicitly linked to the scipy original functions and explain if `diags` is same as `spdiags`

Also, if diagonals is a 1-dim tensor, it would be awesome if `offsets = torch.tensor([0])` was default or if at least `offsets = 0` was supported

**Comment by vadimkantorov** [comment]
Btw scipy supports not passing `offsets` at all and uses 0 offset by default
